pkg/test/extensions/binary.go

```diff
@@ -15,71 +16,151 @@ import (
 func (b *TestBinary) ListTests(ctx context.Context) (ExtensionTestSpecs, error) {
 	var tests ExtensionTestSpecs
 	start := time.Now()
-	binName := filepath.Base(b.path)
+	binName := filepath.Base(b.binaryPath)
 
-	b.logger.Printf("Listing tests for %q", binName)
-	command := exec.Command(b.path, "list")
+	fmt.Fprintf(os.Stderr, "Listing tests for %q", binName)
+	command := exec.Command(b.binaryPath, "list", "-o", "jsonl")
 	testList, err := runWithTimeout(ctx, command, 10*time.Minute)
 	if err != nil {
-		return nil, fmt.Errorf("failed running '%s list': %w", b.path, err)
+		return nil, fmt.Errorf("failed running '%s list': %w", b.binaryPath, err)
 	}
 	buf := bytes.NewBuffer(testList)
 	for {
 		line, err := buf.ReadString('\n')
 		if err == io.EOF {
 			break
 		}
-		if !strings.HasPrefix(line, "[{") {
+		if !strings.HasPrefix(line, "{") {
 			continue
 		}
 
-		var extensionTestSpecs ExtensionTestSpecs
-		err = json.Unmarshal([]byte(line), &extensionTestSpecs)
+		extensionTestSpec := new(ExtensionTestSpec)
+		err = json.Unmarshal([]byte(line), extensionTestSpec)
 		if err != nil {
-			return nil, err
-		}
-		for i := range extensionTestSpecs {
-			extensionTestSpecs[i].Binary = b.path
+			return nil, errors.Wrapf(err, "line: %s", line)
 		}
-		tests = append(tests, extensionTestSpecs...)
+		extensionTestSpec.Binary = b
+		tests = append(tests, extensionTestSpec)
 	}
-	b.logger.Printf("Listed %d tests for %q in %v", len(tests), binName, time.Since(start))
+	fmt.Fprintf(os.Stderr, "Listed %d tests for %q in %v", len(tests), binName, time.Since(start))
 	return tests, nil
 }
 
+// RunTests executes the named tests and returns the results.
+func (b *TestBinary) RunTests(ctx context.Context, timeout time.Duration, env []string,
+	names ...string) []*ExtensionTestResult {
+	var results []*ExtensionTestResult
+	unseenTests := sets.New[string](names...)
+	binName := filepath.Base(b.binaryPath)
+
+	// Build command
+	args := []string{"run-test"}
+	for _, name := range names {
+		args = append(args, "-n", name)
+	}
+	args = append(args, "-o", "jsonl")
+	command := exec.Command(b.binaryPath, args...)
+	if len(env) == 0 {
+		env = os.Environ()
+	}
+	command.Env = env
+
+	// Run test
+	testResult, err := runWithTimeout(ctx, command, timeout)
+	buf := bytes.NewBuffer(testResult)
+	for {
+		line, err := buf.ReadString('\n')
+		if err == io.EOF {
+			break
+		}
+		if !strings.HasPrefix(line, "{") {
+			continue
+		}
+		result := new(ExtensionTestResult)
+		err = json.Unmarshal([]byte(line), &result)
+		if err != nil {
+			panic(fmt.Sprintf("test binary %q returned unmarshallable result", binName))
+		}
+		if !unseenTests.Has(result.Name) {
+			result.Result = ResultFailed
+			result.Error = fmt.Sprintf("test binary %q returned unexpected result: %s", binName, result.Name)
+		}
+		unseenTests.Delete(result.Name)
+		results = append(results, result)
+	}
+	if len(results) == 0 && err != nil {
```

Comment:
would it make more sense to check for the error prior to (and instead of) the for loop that reads the lines? Is there any case where we will want to read the `testResult` when the command has errored?

Reply:
No, extension binaries return non-zero if any tests failed, but that's not the error we're worried about here.  We're only worried if it returned 0 and also didn't give us any results back, does that make sense?

Reply:
I probably don't need this, I wanted to differentiate between not getting any results vs not getting some results, but I guess it's not important.

Reply:
OTE could have a convention about exit codes instead, where we differentiate between failed tests and some other error occurring, WDYT?

Reply:
> OTE could have a convention about exit codes instead, where we differentiate between failed tests and some other error occurring, WDYT?

That definitely seems useful to me